### ISSUE TITLE
Matmul hangs when passed ndarray with dtype np.float16

### ISSUE THREAD

**Issue by wkyle** [numpy/numpy#11590]
Using numpy version 1.14.5 with openBLAS. Passing two arrays of half-precision floats to matmul causes CPU usage to climb to 100% and the computation never finishes. Single and double precision arrays work fine. This failure also occurs using Tensorflow.matmul with np.float16 arrays. I read that BLAS can only work with single and double precision and complex arrays so perhaps numpy is failing to fall back to its default multiplication routine when given half-precision arrays.

To reproduce:
```
import numpy as np
A = np.random.rand(1000, 1000).astype(np.float16)

# This will hang
B = np.matmul(A, A.T)
```

Let me know if I can provide any further information about the conditions surrounding this issue.

**Comment by charris** [comment]
Works for me on linux, python 3.6, NumPy master. What platform and where did you get NumPy?

**Comment by seberg** [comment]
That seems strange, I would think there is no interaction with blas at all for float16?

**Comment by seberg** [comment]
@wkyle are you sure it is hanging as opposed to just being increadibly slow? Timing it gives me a factor of 10 slower for the float16 version compared to float32, which probably has to be expected. (which on my comp means 10 seconds for the computation)!

**Comment by wkyle** [comment]
@seberg I did further tests and you are right. It is just extremely slow. I didn't expect the non-BLAS routine to be so dramatically different.

**Comment by seberg** [comment]
Well, a factor of 10 for BLAS compared to naïve implementations is
probably normal. Wouldn't have been too surprised if it was more then
that due to constantly casting float16 to float32.

Anyway, one last remark you are probably aware of, but one should be
careful already with float32 math, so I would avoid float16 math unless
 you know well that it will be OK for your application (it might be the
matmul/dot operation is summing in float32 here, but I do not trust it.
More importantly I would not just trust it to always sum in float32
here when the memory order changes unless you test it/check the code).

Long story short: Avoid the float16 (or even float32) unless you are
pretty sure it is fine (e.g. I use float32 for video analysis, but
would never use it for simulations).
